### ISSUE TITLE
OpenSSH functionality broken for Windows users on 6.1.0+

### ISSUE THREAD

**Issue by Aharic** [gravitational/teleport#6955]
### Description

**What happened**:

When attempting to connect with OpenSSH on Windows through the Teleport proxy acting as JumpProxy while running tsh 6.1.0+, Proxy kicks back a `permission denied (public key)` error. This works without issue on pre 6.1.0 versions including 6.0.3 (latest available for testing).

**What you expected to happen**:

OpenSSH integration should continue to work on 6.1.0+ for tsh clients running on Windows OS.

### Reproduction Steps

SSH Config:

```
Host examplehost
    User unixuser
    Port 3022
    IdentityFile C:\Users\User\.tsh\keys\teleport.example.com\exampleuser
    ProxyCommand ssh.exe -i C:\Users\User\.tsh\keys\teleport.example.com\exampleuser unixuser@teleport.example.com -p 3023 -s proxy:%h:%p"
```

1. Log into Teleport cluster w/ `tsh login`
2. Confirm keys exist in location specified by IdentityFile
3. run `ssh examplehost`
4. Observe `permission denied (public key)` error

Same happens when attempting to connect directly to the proxy node with `ssh.exe -i C:\Users\User\.tsh\keys\teleport.example.com\exampleuser unixuser@proxy.example.com -p 3022
`

### Server Details
- Teleport version (run `teleport version`): 6.1.0+

### Client Details
- Tsh version (`tsh version`): 6.1.0+
- Computer OS (e.g. Linux, macOS, Windows): Windows

**Comment by jdconti** [comment]
I can confirm this behavior as well, this breaks difficult ssh clients like MobaXterm which don't support the native OpenSSH ssh-agent.

**Comment by benarent** [comment]
@timothyb89 This looks like a regression, @timothyb89 Can you PTAL since you're currently working on the OpenSSH configuration helper.

**Comment by timothyb89** [comment]
I can reproduce this on my macOS box as long as the SSH agent is disabled (`unset SSH_AUTH_SOCK`). It doesn't depend on server version, either - I think this is a quirk of how `ssh` loads certificate files.

On v6.0.3, tsh's `keys/` directory has a layout like this:
```bash
$ ll ~/.tsh-v6.0.3/keys/proxy.foo.example.com/
drwx------    - tim 29 Jun 15:11 /Users/tim/.tsh-v6.0.3/keys/proxy.foo.example.com/
.rw-r----- 2.6k tim 29 Jun 15:11 ├── certs.pem
.rw------- 1.7k tim 29 Jun 15:11 ├── tim
.rw------- 1.7k tim 29 Jun 15:11 ├── tim-cert.pub
drwx------    - tim 29 Jun 15:11 ├── tim-kube/
drwx------    - tim 29 Jun 15:11 │  └── foo.example.com/
.rw------- 1.5k tim 29 Jun 15:11 ├── tim-x509.pem
.rw-------  381 tim 29 Jun 15:11 └── tim.pub
```

On v6.1.5, this changes slightly:
```bash
$ ll ~/.tsh-v6.1.5/keys/proxy.foo.example.com/
drwx------    - tim 29 Jun 15:09 /Users/tim/.tsh-v6.1.5/keys/proxy.foo.example.com/
.rw-r----- 2.6k tim 29 Jun 15:09 ├── certs.pem
.rw------- 1.7k tim 29 Jun 15:09 ├── tim
drwx------    - tim 29 Jun 15:09 ├── tim-ssh/
.rw------- 1.7k tim 29 Jun 15:09 │  └── foo.example.com-cert.pub
.rw------- 1.5k tim 29 Jun 15:09 ├── tim-x509.pem
.rw-------  381 tim 29 Jun 15:09 └── tim.pub
```

In v6.0.3, the private key - in my case `~/.tsh/keys/proxy.foo.example.com/tim` - had its associated certificate file colocated in the same directory, at `tim-cert.pub`. `ssh` loads these certificate files automatically, which you can (kind of) see in the debug output:
```bash
$ ssh -v node000 exit 0 2>&1 | grep -i cert
debug1: identity file /Users/tim/.tsh/keys/proxy.foo.example.com/tim-cert type 4
debug1: kex: host key algorithm: ssh-rsa-cert-v01@openssh.com
debug1: Server host certificate: ssh-rsa-cert-v01@openssh.com SHA256:XxfU+Zp077pcyIAKJg3aA1gqt42rx3O7C6ugdIpZ5VA, serial 0 ID "" CA ssh-rsa SHA256:Y5svvbvZKBxzI/wPK/JE0z3/Bp6T3R5JW10peuztKjs valid after 2021-06-11T16:54:11
debug1: Will attempt key: /Users/tim/.tsh/keys/proxy.foo.example.com/tim RSA-CERT SHA256:/hGHPSPlrvISr/6NdHRYStvzbUl7T7Gt/X+4lliWyQM explicit
debug1: Offering public key: /Users/tim/.tsh/keys/proxy.foo.example.com/tim RSA-CERT SHA256:/hGHPSPlrvISr/6NdHRYStvzbUl7T7Gt/X+4lliWyQM explicit
debug1: Server accepts key: /Users/tim/.tsh/keys/proxy.foo.example.com/tim RSA-CERT SHA256:/hGHPSPlrvISr/6NdHRYStvzbUl7T7Gt/X+4lliWyQM explicit
```

In newer versions of Teleport, this file has been moved to `[user]-ssh/[clusterName]-cert.pub`. If we try that same command, we can see ssh fails to automatically find the certificate:
```bash
$ ssh -v node000 exit 0 2>&1 | grep -i cert
debug1: identity file /Users/tim/.tsh/keys/proxy.foo.example.com/tim-cert type -1
```

From the server's perspective, it sees an `ssh-rsa` auth attempt like this on failure:
```
DEBU [PROXY]     conn(127.0.0.1:61162->127.0.0.1:3023, user=tim) auth attempt fingerprint:ssh-rsa SHA256:LlJd9P02SvzyN6xXoHFbyWdPf06sy7HCPkQ7d/r0rmE local:127.0.0.1:3023 remote:127.0.0.1:61162 user:tim srv/authhandlers.go:212
DEBU [PROXY]     conn(127.0.0.1:61162->127.0.0.1:3023, user=tim) auth attempt with key ssh-rsa SHA256:LlJd9P02SvzyN6xXoHFbyWdPf06sy7HCPkQ7d/r0rmE, (*ssh.Certificate)(nil) fingerprint:ssh-rsa SHA256:LlJd9P02SvzyN6xXoHFbyWdPf06sy7HCPkQ7d/r0rmE local:127.0.0.1:3023 remote:127.0.0.1:61162 user:tim srv/authhandlers.go:215
DEBU [PROXY]     auth attempt, unsupported key type fingerprint:ssh-rsa SHA256:LlJd9P02SvzyN6xXoHFbyWdPf06sy7HCPkQ7d/r0rmE local:127.0.0.1:3023 remote:127.0.0.1:61162 user:tim srv/authhandlers.go:217
```

...and this on success:
```
DEBU [PROXY]     conn(127.0.0.1:61179->127.0.0.1:3023, user=tim) auth attempt fingerprint:ssh-rsa SHA256:/hGHPSPlrvISr/6NdHRYStvzbUl7T7Gt/X+4lliWyQM local:127.0.0.1:3023 remote:127.0.0.1:61179 user:tim srv/authhandlers.go:212
DEBU [PROXY]     conn(127.0.0.1:61179->127.0.0.1:3023, user=tim) auth attempt with key ssh-rsa SHA256:/hGHPSPlrvISr/6NdHRYStvzbUl7T7Gt/X+4lliWyQM, (*ssh.Certificate)(nil) fingerprint:ssh-rsa SHA256:/hGHPSPlrvISr/6NdHRYStvzbUl7T7Gt/X+4lliWyQM local:127.0.0.1:3023 remote:127.0.0.1:61179 user:tim srv/authhandlers.go:215
DEBU [PROXY]     auth attempt, unsupported key type fingerprint:ssh-rsa SHA256:/hGHPSPlrvISr/6NdHRYStvzbUl7T7Gt/X+4lliWyQM local:127.0.0.1:3023 remote:127.0.0.1:61179 user:tim srv/authhandlers.go:217
DEBU [PROXY]     conn(127.0.0.1:61179->127.0.0.1:3023, user=tim) auth attempt fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:CggfPWSDW4Dhal/VwQzvFnvvX14Z90lnTNHq2yVwl5k local:127.0.0.1:3023 remote:127.0.0.1:61179 user:tim srv/authhandlers.go:212
DEBU [PROXY]     conn(127.0.0.1:61179->127.0.0.1:3023, user=tim) auth attempt with key ssh-rsa-cert-v01@openssh.com SHA256:CggfPWSDW4Dhal/VwQzvFnvvX14Z90lnTNHq2yVwl5k, &ssh.Certificate{Nonce:[]uint8{0x2c, 0xfb, 0xf5, 0x82, 0xba, 0xa3, 0x1f, 0xc0, 0xb, 0xf1, 0xfb, 0x75, 0xc9, 0xea, 0x29, 0xb9, 0x46, 0xef, 0x26, 0x67, 0x86, 0x3a, 0x1e, 0x8b, 0xac, 0xee, 0x29, 0x50, 0xc6, 0x80, 0x86, 0x2e}, Key:(*ssh.rsaPublicKey)(0xc001334a50), Serial:0x0, CertType:0x1, KeyId:"tim", ValidPrincipals:[]string{"root", "tim"}, ValidAfter:0x60db8c29, ValidBefore:0x60dc3525, Permissions:ssh.Permissions{CriticalOptions:map[string]string{}, Extensions:map[string]string{"permit-agent-forwarding":"", "permit-port-forwarding":"", "permit-pty":"", "teleport-roles":"{\"version\":\"v1\",\"roles\":[\"access\",\"editor\",\"admin\"]}", "teleport-route-to-cluster":"foo.example.com", "teleport-traits":"{\"kubernetes_groups\":[\"\"],\"kubernetes_users\":[\"\"],\"logins\":[\"root\",\"tim\"]}"}}, Reserved:[]uint8{}, SignatureKey:(*ssh.rsaPublicKey)(0xc001334a90), Signature:(*ssh.Signature)(0xc000f81f00)} fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:CggfPWSDW4Dhal/VwQzvFnvvX14Z90lnTNHq2yVwl5k local:127.0.0.1:3023 remote:127.0.0.1:61179 user:tim srv/authhandlers.go:215
DEBU [PROXY]     Successfully authenticated fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:CggfPWSDW4Dhal/VwQzvFnvvX14Z90lnTNHq2yVwl5k local:127.0.0.1:3023 remote:127.0.0.1:61179 user:tim srv/authhandlers.go:275
```

A relatively easy workaround is tell `ssh` exactly where to find your certificate file. I'm partial to this config snippet which is similar to what we'll be automatically generating in #7437:

```
# Rules for all Teleport cluster hosts (including the proxy)
Host *.foo.example.com proxy.foo.example.com
    IdentityFile "/Users/tim/.tsh/keys/proxy.foo.example.com/tim"
    CertificateFile "/Users/tim/.tsh/keys/proxy.foo.example.com/tim-ssh/foo.example.com-cert.pub"

# Rules for only Teleport cluster nodes (excluding the proxy)
Host *.foo.example.com !proxy.foo.example.com
    Port 3022
    ProxyCommand ssh -p 3023 proxy.foo.example.com -s proxy:%h:%p
```

... which will allow `ssh node.foo.example.com`, but to adapt the example from the ticket, something like this should also work fine:

```
Host node000
    User tim
    Port 3022
    IdentityFile /Users/tim/.tsh/keys/proxy.foo.example.com/tim
    CertificateFile "/Users/tim/.tsh/keys/proxy.foo.example.com/tim-ssh/foo.example.com-cert.pub"
    ProxyCommand ssh -o "CertificateFile ~/.tsh/keys/proxy.foo.example.com/tim-ssh/foo.example.com-cert.pub" -i ~/.tsh/keys/proxy.foo.example.com/tim tim@proxy.foo.example.com -p 3023 -s proxy:%h:%p
```

I'll have to defer elsewhere as to whether or not this should be treated as a regression. I believe the change to the SSH certificate path was introduced in https://github.com/gravitational/teleport/pull/5938.

**Comment by timothyb89** [comment]
Small update on this, the upcoming patch to add Windows support to `tsh config` will generate configuration like I described above automatically: https://github.com/gravitational/teleport/pull/7848

**Comment by zmb3** [comment]
Looks like this was fixed with #7848
